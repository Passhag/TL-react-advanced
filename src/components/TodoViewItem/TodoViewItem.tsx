import * as React from 'react';
import { Todo } from '../../api';

interface Props {
  item: Todo;
  onDeleteBtnClick(event: React.MouseEvent<HTMLButtonElement>): void;
  onEditBtnClick(event: React.MouseEvent<HTMLButtonElement>): void;
}

const TodoViewItem: React.SFC<Props> = ({ item, onDeleteBtnClick, onEditBtnClick }): JSX.Element => {
  return (
    <div className="tl-todo-item">
      <span className="tl-todo-item-text">
        {item.content}
      </span>
      <div className="tl-todo-item-actions">
        <button type="button" className="btn btn-link" onClick={onEditBtnClick}>
          <i className="fa fa-pencil" />
        </button>
        <button type="button" className="btn btn-link" onClick={onDeleteBtnClick}>
          <i className="fa fa-trash" />
        </button>
      </div>
    </div>
  );
};

export default TodoViewItem;
