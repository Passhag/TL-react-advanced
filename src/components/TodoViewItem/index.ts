import TodoViewItem from './TodoViewItem';

export {
  TodoViewItem,
};
