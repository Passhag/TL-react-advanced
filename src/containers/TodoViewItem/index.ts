import TodoViewItemContainer from './TodoViewItem';

export {
  TodoViewItemContainer,
};
