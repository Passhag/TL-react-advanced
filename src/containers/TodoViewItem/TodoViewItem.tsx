import * as React from 'react';
import { connect } from 'react-redux';
import { removeTodo, updateTodo, setIsOpen, setContent } from '../../redux/actions';
import { Todo } from '../../api';
import { TodoViewItem } from '../../components/TodoViewItem';

const ContentCmp = () => <h1>Modal Content</h1>;

interface Props {
  item: Todo & { id: string };
  removeTodo(id: string): () => void;
  updateTodo(todo: Todo): () => void;
  openModal(): any;
  setModalContent(arg: any): any;
}

class TodoViewItemContainer extends React.Component<Props, {}> {
  constructor(props: any) {
    super(props);

    this.handleDeleteBtnOnClick = this.handleDeleteBtnOnClick.bind(this);
    this.handleEditBtnOnClick = this.handleEditBtnOnClick.bind(this);
  }

  handleDeleteBtnOnClick(event: React.MouseEvent<HTMLButtonElement>): void {
    event.stopPropagation();

    this.props.removeTodo(this.props.item.id);
  }

  handleEditBtnOnClick(event: React.MouseEvent<HTMLButtonElement>): void {
    event.stopPropagation();
    this.props.setModalContent(ContentCmp);
    this.props.openModal();
  }

  render(): JSX.Element {
    return (
      <TodoViewItem
        item={this.props.item}
        onDeleteBtnClick={this.handleDeleteBtnOnClick}
        onEditBtnClick={this.handleEditBtnOnClick}
      />
    );
  }
}

const mapDispatchToProps = (dispatch: Function) => ({
  removeTodo: (id: string) => dispatch(removeTodo(id)),
  updateTodo: (todo: Todo) => dispatch(updateTodo(todo)),
  openModal: () => dispatch(setIsOpen(true)),
  setModalContent: (ContentCmp) => dispatch(setContent(ContentCmp)),
});

export default connect(undefined, mapDispatchToProps)(TodoViewItemContainer);
