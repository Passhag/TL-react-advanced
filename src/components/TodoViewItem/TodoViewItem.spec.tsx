import * as React from 'react';
import { shallow, configure, ShallowWrapper } from 'enzyme';
import * as ReactSixteenAdapter from 'enzyme-adapter-react-16';
import TodoViewItem from './TodoViewItem';

configure({ adapter: new ReactSixteenAdapter() });

const aTodo = () => ({
  content: 'todo',
});

const aShallowWrapper = (options: { [key: string]: Function } = {}): ShallowWrapper => {
  const props = Object.assign({}, {
    item: aTodo(),
    onDeleteBtnClick: () => null,
    onEditBtnClick: () => null,
  }, options);

  return shallow(
    <TodoViewItem {...props} />
  );
};

describe('Components::TodoViewItem', () => {
  it('should render correctly', () => {
    const wrapper = aShallowWrapper();

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle onDeleteBtnClick', () => {
    const onDeleteBtnClick = jasmine.createSpy('onDeleteBtnClick');
    const wrapper = aShallowWrapper({ onDeleteBtnClick, });

    wrapper.find('button').simulate('click');

    expect(onDeleteBtnClick).toHaveBeenCalled();
  });

  it('should handle onEditBtnClick', () => {
    const onEditBtnClick = jasmine.createSpy('onEditBtnClick');
    const wrapper = aShallowWrapper({ onEditBtnClick, });

    wrapper.find('button').simulate('click');

    expect(onEditBtnClick).toHaveBeenCalled();
  });
});
