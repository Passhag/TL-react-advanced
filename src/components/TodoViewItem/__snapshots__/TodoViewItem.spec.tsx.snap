// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components::TodoViewItem should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoViewItem
    item={
        Object {
            "content": "todo",
          }
    }
    onDeleteBtnClick={[Function]}
    onEditBtnClick={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <span
          className="tl-todo-item-text"
>
          todo
</span>,
        <div
          className="tl-todo-item-actions"
>
          <button
                    className="btn btn-link"
                    onClick={[Function]}
                    type="button"
          >
                    <i
                              className="fa fa-pencil"
                    />
          </button>
          <button
                    className="btn btn-link"
                    onClick={[Function]}
                    type="button"
          >
                    <i
                              className="fa fa-trash"
                    />
          </button>
</div>,
      ],
      "className": "tl-todo-item",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "todo",
          "className": "tl-todo-item-text",
        },
        "ref": null,
        "rendered": "todo",
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="btn btn-link"
              onClick={[Function]}
              type="button"
>
              <i
                            className="fa fa-pencil"
              />
</button>,
            <button
              className="btn btn-link"
              onClick={[Function]}
              type="button"
>
              <i
                            className="fa fa-trash"
              />
</button>,
          ],
          "className": "tl-todo-item-actions",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <i
                className="fa fa-pencil"
/>,
              "className": "btn btn-link",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "fa fa-pencil",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <i
                className="fa fa-trash"
/>,
              "className": "btn btn-link",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "fa fa-trash",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <span
            className="tl-todo-item-text"
>
            todo
</span>,
          <div
            className="tl-todo-item-actions"
>
            <button
                        className="btn btn-link"
                        onClick={[Function]}
                        type="button"
            >
                        <i
                                    className="fa fa-pencil"
                        />
            </button>
            <button
                        className="btn btn-link"
                        onClick={[Function]}
                        type="button"
            >
                        <i
                                    className="fa fa-trash"
                        />
            </button>
</div>,
        ],
        "className": "tl-todo-item",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "todo",
            "className": "tl-todo-item-text",
          },
          "ref": null,
          "rendered": "todo",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="btn btn-link"
                onClick={[Function]}
                type="button"
>
                <i
                                className="fa fa-pencil"
                />
</button>,
              <button
                className="btn btn-link"
                onClick={[Function]}
                type="button"
>
                <i
                                className="fa fa-trash"
                />
</button>,
            ],
            "className": "tl-todo-item-actions",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <i
                  className="fa fa-pencil"
/>,
                "className": "btn btn-link",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "fa fa-pencil",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <i
                  className="fa fa-trash"
/>,
                "className": "btn btn-link",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "fa fa-trash",
                },
                "ref": null,
                "rendered": null,
                "type": "i",
              },
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
